### PR TITLE
Changes to remove errors for high degree polynomials

### LINKED ISSUE #25736: ReflectometryBackgroundSubtraction - fitting wrong for high degree polynomials
In `ReflectometryBackgroundSubtraction`, if the `BackgroundCalculationMethod` is set to `Polynomial` and the `degreeOfPolynomial` is set to a high number the resulting background is incorrect. This error comes from the child algorithm `CalculatePolynomialBackground`. Once it has fit the polynomial the parameters are converted to a string before creating the function which creates large errors for high polynomials. Directly passing the parameters should reduce this error.

If the following script is run it should output a workspace with all zeros with a peak of 5 in the 9th spectra but it doesn't (if the degree of polynomial is lower than 10 it produces the correct values)
```
dataX = [1]
polynomial = [1, 4.75, 8, 10.75, 13, 14.75, 16, 16.75, 17, 16.75, 16, 14.75, 13, 10.75, 8, 4.75, 1] 
peak = [0, 0, 0, 0, 0, 0, 0, 0, 5, 0, 0, 0, 0, 0, 0, 0, 0]
dataY = [a + b for a, b in zip(polynomial, peak)]
ws = CreateWorkspace(dataX, dataY, NSpec = 17)
ws_bkg_subtr = ReflectometryBackgroundSubtraction(ws, InputWorkspaceIndexType='SpectrumNumber', InputWorkspaceIndexSet = '1-8,10-17', BackgroundCalculationMethod = 'Polynomial', DegreeOfPolynomial = 15)
```

### PR DESCRIPTION
**Pull request by alicerussell1**
**Description of work.**
This PR removes the errors found when trying to fit a high degree polynomial to the background. These changes were made in it's child algorithm `CalculatePolynomialBackground`.
The parameters of the polynomial found by the fitting function have been changed to be passed directly rather than being converted to a string.
An additional property has been added to `CalculatePolynomialBackground` to allow the user to change the minimizer used by the fitting function to  `Levenberg-Marquardt`. 

**To test:**
Run the following script, it should output a workspace which contains all zeros except for a peak in the 9th spectra (with some small errors)
```
dataX = [1]
polynomial = [1, 4.75, 8, 10.75, 13, 14.75, 16, 16.75, 17, 16.75, 16, 14.75, 13, 10.75, 8, 4.75, 1] 
peak = [0, 0, 0, 0, 0, 0, 0, 0, 5, 0, 0, 0, 0, 0, 0, 0, 0]
dataY = [a + b for a, b in zip(polynomial, peak)]
ws = CreateWorkspace(dataX, dataY, NSpec = 17)
ws_bkg_subtr = ReflectometryBackgroundSubtraction(ws, InputWorkspaceIndexType='SpectrumNumber', ProcessingInstructions= '1-8,10-17', BackgroundCalculationMethod = 'Polynomial', DegreeOfPolynomial = 15)
```
Fixes #25736 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/Algorithms/src/CalculatePolynomialBackground.cpp [hunk 1]
```diff
@@ -173,7 +180,8 @@ std::vector<double> invertRanges(const std::vector<double> &ranges) {
 std::vector<double>
 executeFit(Mantid::API::Algorithm &fit, const std::string &function,
            Mantid::API::MatrixWorkspace_sptr &ws, const size_t wsIndex,
-           const std::vector<double> &ranges, const std::string &costFunction) {
+           const std::vector<double> &ranges, const std::string &costFunction,
```
Comment (gvardany) [inline Framework/Algorithms/src/CalculatePolynomialBackground.cpp]: Can you add the oxygen for the new arguments?

File: Framework/Algorithms/src/CalculatePolynomialBackground.cpp [hunk 2]
```diff
@@ -199,38 +207,61 @@ executeFit(Mantid::API::Algorithm &fit, const std::string &function,
  *  @param parameters a vector containing the polynomial coefficients
  *  @return a function string
  */
-std::string makeFunctionString(const std::vector<double> &parameters) {
+std::string makeFunctionString(const std::string &name,
```
Comment (gvardany) [inline Framework/Algorithms/src/CalculatePolynomialBackground.cpp]: Same here.

File: Framework/Algorithms/src/CalculatePolynomialBackground.cpp [hunk 3]
```diff
@@ -199,38 +207,61 @@ executeFit(Mantid::API::Algorithm &fit, const std::string &function,
...
+                               const std::vector<double> &parameters) {
   const auto degree = parameters.size() - 1;
```
Comment (gvardany) [inline Framework/Algorithms/src/CalculatePolynomialBackground.cpp]: Is there any guarantee somewhere that parameters size is at least 1?

File: Framework/Algorithms/src/CalculatePolynomialBackground.cpp [hunk 4]
```diff
@@ -199,38 +207,61 @@ executeFit(Mantid::API::Algorithm &fit, const std::string &function,
...
+  if (degree > 2)
+    s << ",n=" << degree;
+  for (size_t d = 0; d <= degree; ++d) {
+    s << ',' << 'A' << d << '=' << parameters[d];
+  }
+  return s.str();
+}
+
+/** Return a name of the function used in the fit.
+ *  @param parameters a vector containing the polynomial coefficients
```
Comment (gvardany) [inline Framework/Algorithms/src/CalculatePolynomialBackground.cpp]: degree?

File: Framework/Algorithms/src/CalculatePolynomialBackground.cpp [hunk 5]
```diff
@@ -199,38 +207,61 @@ executeFit(Mantid::API::Algorithm &fit, const std::string &function,
...
+  return name.str();
 }
 
 /** Evaluates the given function directly on a histogram
  *  @param function a string representing function to evaluate
  *  @param ws an output workspace
  *  @param wsIndex a workspace index identifying a histogram
  */
-void evaluateInPlace(const std::string &function,
+void evaluateInPlace(const std::string &name,
```
Comment (gvardany) [inline Framework/Algorithms/src/CalculatePolynomialBackground.cpp]: again doxygen needs an update

File: Framework/Algorithms/src/CalculatePolynomialBackground.cpp [hunk 6]
```diff
@@ -290,6 +321,12 @@ void CalculatePolynomialBackground::init() {
       Prop::COST_FUNCTION, CostFunc::WEIGHTED_LEAST_SQUARES.c_str(),
       boost::make_shared<Kernel::ListValidator<std::string>>(costFuncOpts),
       "The cost function to be passed to the Fit algorithm.");
+  std::array<std::string, 2> minimizerOpts{
+      {Minimizer::LEVENBERG_MARQUARDT_MD, Minimizer::LEVENBERG_MARQUARDT}};
+  declareProperty(
+      Prop::MINIMIZER, Minimizer::LEVENBERG_MARQUARDT_MD.c_str(),
```
Comment (gvardany) [inline Framework/Algorithms/src/CalculatePolynomialBackground.cpp]: not sure you need c_str() here

File: Framework/Algorithms/src/ReflectometryBackgroundSubtraction.cpp [hunk 7]
```diff
@@ -270,6 +271,14 @@ void ReflectometryBackgroundSubtraction::exec() {
   }
 
   if (backgroundType == "Polynomial") {
+    int range = static_cast<int>(spectraList.back() - spectraList.front());
+    int degree = getProperty("degreeOfPolynomial");
+    if (range < degree) {
```
Comment (gvardany) [inline Framework/Algorithms/src/ReflectometryBackgroundSubtraction.cpp]: what if range == degree ?

File: Testing/SystemTests/tests/analysis/ILLDirectGeometryReductionTest.py [hunk 8]
```diff
@@ -145,7 +145,7 @@ def runTest(self):
         mtd['cropped'].mutableRun().addProperty('run_title', '', True)
 
     def validate(self):
-        self.tolerance = 1e-7
+        self.tolerance = 1e-3
```
Comment (gvardany) [inline Testing/SystemTests/tests/analysis/ILLDirectGeometryReductionTest.py]: This is a dramatic change in tolerance. Maybe updating the reference and keeping the tolerance is a better idea. Can you post a comparison of the output of this test with the current reference?

**Review by gvardany — COMMENTED** [review]
Please see below:

File: Framework/Algorithms/src/CalculatePolynomialBackground.cpp (hunk 3 above)
Comment (alicerussell1) [reply Framework/Algorithms/src/CalculatePolynomialBackground.cpp]: Yes, in exec the size of parameters is degree + 1 and there is validation to check degree is > 0

File: Framework/Algorithms/src/ReflectometryBackgroundSubtraction.cpp (hunk 7 above)
Comment (alicerussell1) [reply Framework/Algorithms/src/ReflectometryBackgroundSubtraction.cpp]: Fitting should work if they are equal as the number of spectra in the list will be 1 greater than the range (e.g if the spectra list is 1, 2 and 3, then the range will be 3 - 1)

File: Framework/Algorithms/src/CalculatePolynomialBackground.cpp [hunk 9]
```diff
@@ -196,41 +206,66 @@ executeFit(Mantid::API::Algorithm &fit, const std::string &function,
...
-  // We want to write directly to the workspace.
+      Mantid::API::FunctionFactory::Instance().createFunction(name));
+  if (degree > 2) {
+    Mantid::API::IFunction1D::Attribute att = bkg->getAttribute("n");
+    att.fromString(std::to_string(degree));
+    bkg->setAttribute("n", att);
+  }
+  for (size_t d = 0; d <= degree; ++d) {
+    std::string param = 'A' + std::to_string(d);
+    bkg->setParameter(param, parameters[d]);
```
Comment (gvardany) [inline Framework/Algorithms/src/CalculatePolynomialBackground.cpp]: I'm trying to wrap my head around why IN5 system test needed to be updated.
Can you confirm that it's this block that is changing the behaviour also for 0-degree fitting?

File: docs/source/release/v4.1.0/framework.rst [hunk 10]
```diff
@@ -39,6 +39,7 @@ Improvements
 - Prevent an error due to the locale settings which may appear when reading, for instance, the incident energy Ei value from the logs in :ref:`ConvertUnits <algm-ConvertUnits>` and many other algorithms.
 - :code:`indices` and :code:`slicepoint` options have been added to :ref:`mantid.plots <mantid.plots>` to allow selection of which plane to plot from an MDHistoWorkspace. :code:`transpose` has also been added to transpose the axes of any 2D plot.
 - :ref:`Pseudo-Voigt <func-PseudoVoigt>` has been modified to be more in line with FULLPROF and GSAS.  One of its basic parameter, Height, is changed to Intensity.
+- :ref:`CalculatePolynomialBackground <algm-CalculatePolynomialBackground>` has an option to select the the minimizer used during fitting.
```
Comment (gvardany) [inline docs/source/release/v4.1.0/framework.rst]: I think here we should also add a note that it's more correct now due to passing parameters around with double precision.

File: Framework/Algorithms/src/CalculatePolynomialBackground.cpp (hunk 9 above)
Comment (alicerussell1) [reply Framework/Algorithms/src/CalculatePolynomialBackground.cpp]: Yes this block is different for all degrees of polynomial. 

The differences between the new output and the original reference for the system test seems to be quite small. I have just tried setting the relative error to false in the system test there is then only one mismatch of data: 
``` 
Data mismatch at cell (hist#,bin#): (267,179)
 Dataset #1 (X,Y,E) = (2.023,4.68025,0.273321)
 Dataset #2 (X,Y,E) = (2.023,4.68025,0.273322)
 Difference (X,Y,E) = (0,1.18704e-07,1.27676e-07)
```

**Review by gvardany — APPROVED** [review]
Confirmed that setting parameters as doubles is more correct than what was before; the default precision of double to string stream turns out to be 6 decimals. The relative difference of the system test result for IN5 is at most 0.0001.

**Comment by alicerussell1** [comment]
The failing test is unrelated

**Pull request merged**